contrib/platform/src/com/sun/jna/platform/win32/Ddeml.java

```diff
@@ -2581,9 +2581,44 @@ public HDDEDATA DdeCreateDataHandle(
      *   <li>{@link #DMLERR_NO_ERROR}</li>
      *   <li>{@link #DMLERR_SYS_ERROR}</li>
      * </ul>
+     *
+     * @deprecated Use {@link #DdeCreateStringHandle(int, Pointer, int)}
      */
+    @Deprecated
     public HSZ DdeCreateStringHandle(int idInst, String psz, int iCodePage);
 
+    /**
+     * Creates a handle that identifies the specified string. A Dynamic
+     * Data Exchange (DDE) client or server application can pass the string
+     * handle as a parameter to other Dynamic Data Exchange Management
+     * Library (DDEML) functions.
+     *
+     * @param idInst The application instance identifier obtained by a previous
+     * call to the {@link #DdeInitialize} function.
+     *
+     * @param psz The null-terminated string for which a handle is to be created.
+     * This string can be up to 255 characters. The reason for this limit is that
+     * DDEML string management functions are implemented using atoms.
+     *
+     * @param iCodePage The code page to be used to render the string. This
+     * value should be either {@link #CP_WINANSI} (the default code page) or
+     * {@link #CP_WINUNICODE}, depending on whether the ANSI or Unicode version of
+     * {@link #DdeInitialize} was called by the client application.
+     *
+     * @return If the function succeeds, the return value is a string handle.
+     *
+     * <p>If the function fails, the return value is 0L.</p>
+     *
+     * <p>The {@link #DdeGetLastError} function can be used to get the error code, which
+     * can be one of the following values:</p>
+     * <ul>
+     *   <li>{@link #DMLERR_INVALIDPARAMETER}</li>
+     *   <li>{@link #DMLERR_NO_ERROR}</li>
+     *   <li>{@link #DMLERR_SYS_ERROR}</li>
+     * </ul>
+     */
+    public HSZ DdeCreateStringHandle(int idInst, Pointer psz, int iCodePage);
```

Comment:
Cannot have the `psz` as a string here as the encoding of the native memory depends on `iCodePage` that the user provide.

Reply:
It would be good to extract this commit into a separate PR. This change looks sane, I would however not bind this as a Pointer, but a byte[]. Writing memory can be expansive and in this case (JMH benchmark): 

```java
public class MemoryBenchmark {
    static {
        System.setProperty("jna.encoding", "UTF-8");
    }

    public static interface TestLibrary extends Library {

        TestLibrary INSTANCE = Native.load(
            "/home/matthias/src/jnalib/build/native-linux-x86-64/libtestlib.so",
            TestLibrary.class);

        Pointer returnPointerArgument(Pointer ptr);
        Pointer returnPointerArgument(byte[] ptr);
    }


    @Benchmark
    public void allocateWithMemory(Blackhole blackhole) {
        Memory memory = new Memory(11);
        memory.setString(0, "Hallo Welt");
        TestLibrary.INSTANCE.returnPointerArgument(memory);
    }

    @Benchmark
    public void allocateWithByteArray(Blackhole blackhole) {
        byte[] memory = Native.toByteArray("Hallo Welt");
        TestLibrary.INSTANCE.returnPointerArgument(memory);
    }

}
```

results in

```
Benchmark                               Mode  Cnt       Score        Error  Units
MemoryBenchmark.allocateWithByteArray  thrpt   25  885908,737 ±  25941,594  ops/s
MemoryBenchmark.allocateWithMemory     thrpt   25  360946,404 ± 131413,519  ops/s
```

The reason is most probably, that instead of only transitioning to native once, transition to native happens at least 3 times in the Memory variant (creation of the memory object, writing the value, call to target function). The bytearray variant only suffers from a memcopy on the native side when the array is made available to native.

There is another benefit: Heap pressure goes up and the risk for the process dying because it runs out of memory lessens (Memory objects are small and thus the heap can hold many of these and in time these can hold huge amounts of native memory).

Reply:
Okay. Sure, let's use byte[], I just used Memory as I didn't know any better.

Should the old binding that takes a String argument exist but be annotated with @Deprecated or removed?
What are the impacts of having several bindings with different arguments types for the native part and the reflection code in JNA. I saw that there is some unit tests that verifies that there are no "duplicated" methods. I guess binding String and byte[] counts as a "duplicated" method.

Reply:
Please keep the original method and deprecate it. Yes there are duplicate method checks in place, if so, this method should be added to the "yes it is expected" list.

Reply:
Ok. I will create a PR for this, probably tomorrow.

Reply:
I've looked into this a bit more and I'm a little bit puzzled if this is the right way to go.

On MSDN ([DdeCreateStringHandleW](https://docs.microsoft.com/en-us/windows/win32/api/ddeml/nf-ddeml-ddecreatestringhandlew) & [DdeCreateStringHandleA](https://docs.microsoft.com/en-us/windows/win32/api/ddeml/nf-ddeml-ddecreatestringhandlea)) the documentation for `iCodePage` reads:
> The code page to be used to render the string. This value should be either CP_WINANSI (the default code page) or CP_WINUNICODE, depending on whether the ANSI or Unicode version of DdeInitialize was called by the client application.

So, I guess that that the `psz` parameter should follow the mapping configured for Ddeml.
What is the right way to determine the version of DdeInitialize called? Is it using `W32APIOptions.DEFAULT_OPTIONS` or something like `Native.getLibraryOptions(Ddeml.class).get(Library.OPTION_FUNCTION_MAPPER)`?

Reply:
I reread again. And there is actually no problem. You quoted the right part. The caller is responsible to pass in the right value for `iCodePage`, but the bindings themselves do what they can.

`Ddeml` loads the DDE library with `W32APIOptions.DEFAULT_OPTIONS`. So these mappings apply:

Java Construct | `w32.ascii` = `false` | `w32.ascii` = `true`
------------ | ------------- | ------------- 
`Ddeml.INSTANCE.DdeInitialize` | `DdeInitializeW`  | `DdeInitializeA`
`Ddeml.INSTANCE.DdeCreateStringHandle` | `DdeCreateStringHandleW`  | `DdeCreateStringHandleA`
`java.lang.String` | `char[]` / `LPCSTR`  | `wchar[]` / `LPCWSTR`

So with `w32.ascii` set to `true`, calls to `DdeCreateStringHandle` are required to pass `CP_WINANSI`. In the default case (`w32.ascii` = `false`) `CP_WINUNICODE` is required.

This is exactly what `DdemlUtil` does:

```
            if(W32APIOptions.DEFAULT_OPTIONS == W32APIOptions.UNICODE_OPTIONS) {
                codePage = Ddeml.CP_WINUNICODE;
            } else {
                codePage = Ddeml.CP_WINANSI;
            }
            Ddeml.HSZ handle = Ddeml.INSTANCE.DdeCreateStringHandle(idInst, value, codePage);
```

So I have to adjust my judgement. This commit would break the code and should be dropped.